File: 0410. Split Array Largest Sum/Solution.java
class Solution {
    public int splitArray(int[] nums, int m) {
        int left = 0, right = 0;
        for (int num : nums) {
            left = Math.max(left, num);
            right += num;
        }
        while (left < right) {
            int mid = left + ((right - left) >> 1);
            if (valid(nums, m, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return right;
    }

    private boolean valid(int[] nums, int m, int sumLimit) {
        int sum = 0, count = 1;
        for (int num : nums) {
            if (sum + num <= sumLimit) {
                sum += num;
            } else {
                sum = num;
                count += 1;
            }
        }
        return count <= m;
    }
}

File: 0410. Split Array Largest Sum/Solution.py
from typing import List


class Solution:
    def splitArray(self, nums: List[int], m: int) -> int:
        def valid(nums: List[int], m: int, sum_limit: int) -> bool:
            accum, cnt = 0, 1
            for e in nums:
                accum += e
                if accum > sum_limit:
                    accum = e
                    cnt += 1
            return cnt <= m

        left, right = max(nums), sum(nums)
        while left < right:
            mid = left + ((right - left) >> 1)
            if valid(nums, m, mid):
                right = mid
            else:
                left = mid + 1
        return left
